app/app.go

```diff
@@ -602,12 +612,114 @@ func (app *RegenApp) BeginBlocker(ctx sdk.Context, req abci.RequestBeginBlock) a
+		validators := app.StakingKeeper.GetAllValidators(ctx)
+		inStakingValidatorSet := map[string]bool{}
+
+		// memoize tendermint valset votes
+		validatorsVoteInfo := map[string]abci.VoteInfo{} // map of consensus address to validator vote
+		for _, voteInfo := range app.votesInfo {
+			validatorsVoteInfo[sdk.ConsAddress(voteInfo.Validator.Address).String()] = voteInfo
+		}
+
+		for _, validator := range validators {
+			tmProtoPk, err := validator.TmConsPublicKey()
+			if err != nil {
+				panic(err)
+			}
+
+			valCons, err := validator.GetConsAddr()
+			if err != nil {
+				panic(err)
+			}
+
+			if validator.Jailed {
+				if _, ok := validatorsVoteInfo[valCons.String()]; ok {
+					updates = append(updates, abci.ValidatorUpdate{
+						PubKey: tmProtoPk,
+						Power:  0,
+					})
+
+					// We need to reset the missed blockcounter & array so that the validator won't be
+					// immediately slashed for downtime after unjail.
+					signInfo, found := app.SlashingKeeper.GetValidatorSigningInfo(ctx, valCons)
+					if found {
+						signInfo.MissedBlocksCounter = 0
+						signInfo.IndexOffset = 0
+						app.SlashingKeeper.SetValidatorSigningInfo(ctx, valCons, signInfo)
+						missedBlocks := app.SlashingKeeper.GetValidatorMissedBlocks(ctx, valCons)
+						for _, mb := range missedBlocks {
+							app.SlashingKeeper.SetValidatorMissedBlockBitArray(ctx, valCons, mb.Index, false)
+						}
+					}
+
+					inStakingValidatorSet[valCons.String()] = true
+				}
+			} else {
+				updates = append(updates, abci.ValidatorUpdate{
+					PubKey: tmProtoPk,
+					Power:  validator.ConsensusPower(sdk.DefaultPowerReduction),
+				})
+
+				inStakingValidatorSet[valCons.String()] = true
+			}
+		}
+
+		// memoize tendermint valset pubkeys
+		validatorsPubkeysMap := make(map[string]crypto.PubKey, len(validatorsPubkeys.Pubkeys)) // map of consensus address to pubkey
+		for _, pubkey := range validatorsPubkeys.Pubkeys {
+			validatorsPubkeysMap[sdk.ConsAddress(pubkey.Address()).String()] = pubkey
+		}
+
+		for _, voteInfo := range app.votesInfo {
+			consAddr := sdk.ConsAddress(voteInfo.Validator.Address).String()
+			if !inStakingValidatorSet[consAddr] {
+
+				pubkey, ok := validatorsPubkeysMap[consAddr]
+				if !ok {
+					panic(fmt.Sprintf("pubkey not found for validator %s", consAddr))
+				}
+
+				tmPk, err := cryptocodec.FromTmPubKeyInterface(pubkey)
+				if err != nil {
+					panic(err)
+				}
+
+				protoPubKey, err := cryptocodec.ToTmProtoPublicKey(tmPk)
+				if err != nil {
+					panic(err)
+				}
+
+				updates = append(updates, abci.ValidatorUpdate{PubKey: protoPubKey, Power: 0})
+			}
+		}
+
+		return abci.ResponseEndBlock{
+			ValidatorUpdates:      updates,
+			ConsensusParamUpdates: resp.ConsensusParamUpdates,
+			Events:                resp.Events,
+		}
 	}
-	resp.ValidatorUpdates = vals
-	resp.Events = append(resp.Events, events...)
+
+	// we don't have EndBlocker in regen modules, so after the upgrade height return the correct validatorset.
```

Comment:
seems like it could be a source of future errors not even to call the EndBlocker for the regen modules. we should probably call but discard the ValidatorUpdates

Reply:
@aaronc Do you think this is something that can be safely handled in a follow-up when we implement this fix on `main`, instead of on this specific release branch?

Reply:
Best to just do it so it doesn't get forgotten but not a deal breaker

Reply:
> @aaronc Do you think this is something that can be safely handled in a follow-up when we implement this fix on `main`, instead of on this specific release branch?

This is already addressed on `main` branch #933. We migrated regen modules to use SDK Module manager.